### PR TITLE
Tokenizer for Commands and Args

### PR DESCRIPTION
**Pull request by TheMkrage**
Replaced the old system of using a scanner to parse args from the command line.  Now, CommandTokenizer tokenizes the input String and allows for easy use of methods such as getCommand(), or getNextArg(). Arguments and commands can be surrounded by quotes or just spaces

Example of each method being used:
```Java
String commandLineInput = "CommandName Argument1 Argument2";
CommandTokenizer tokenizer = new CommandTokenizer(commandLineInput);
// Will print "Command: CommandName" to the log
System.out.println("Command: " + tokenizer.getCommand());
// Will print "Arg: Argument1"
System.out.println("Arg: " + tokenizer.getNextArg());
// Will print "Remaining Args: 1"
System.out.println("Remaining Args: " + tokenizer.getRemainingArgs().size());
// Will print "Has More Args: false"
System.out.println("Has More Args:  " + tokenizer.hasNextArg());
```

If you need to use the arguments in more than one location, make sure to store the value that each method returns. getNextArg() will continue returning the next argument until there are no more left. Then it will return nill. For example, if you need to use the 1st argument multiple times, be sure to:
```
String arg1 = tokenizer.getNextArg();
```

Reference CommandTokenizerTest.Java for more tests using the tokenizer.

### CODE REVIEW

File: src/codeu/chat/client/commandline/Chat.java [hunk 1]
```diff
@@ -155,8 +154,8 @@ public void invoke(Scanner args) {
     //
     panel.register("u-sign-in", new Panel.Command() {
       @Override
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(List<String> args) {
+    	final String name = !args.isEmpty() ? args.get(0).trim(): "";
```
Comment (julezjw) [inline src/codeu/chat/client/commandline/Chat.java]: nice efficient code here!

File: src/codeu/chat/util/CommandTokenizer.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package codeu.chat.util;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommandTokenizer {
```
Comment (julezjw) [inline src/codeu/chat/util/CommandTokenizer.java]: This is a small detail but do we want to declare this class "final"?

File: src/codeu/chat/util/CommandTokenizer.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	private String source;
+	private int currentIndex = 0;
+
+	public CommandTokenizer(String source) {
```
Comment (julezjw) [inline src/codeu/chat/util/CommandTokenizer.java]: this is another small detail, but for readability purposes, I think it'd be nice to put this function after getCommand() so it makes sense reading this file from top to bottom

**Comment by KevinWorkman** [comment]
Just a high-level comment, but I'd add more info to the pull request description if I were you. Specifically, and example showing how to use this new feature would be great.

File: src/codeu/chat/util/CommandTokenizer.java (hunk 2 above)
Comment (Priya411) [reply src/codeu/chat/util/CommandTokenizer.java]: That's a good point. I declared the new classes I wrote as being final too.

Comment (TheMkrage) [reply src/codeu/chat/util/CommandTokenizer.java]: Making the class final as there is nowhere where this class can be subclassed.

File: src/codeu/chat/client/commandline/Chat.java (hunk 1 above)
Comment (TheMkrage) [reply src/codeu/chat/client/commandline/Chat.java]: Thanks :D

File: src/codeu/chat/util/CommandTokenizer.java (hunk 3 above)
Comment (TheMkrage) [reply src/codeu/chat/util/CommandTokenizer.java]: Changing the order of the methods so that most methods on use methods above themselves. Not sure if there is a proper place to declare how the methods are order.

File: src/codeu/chat/util/CommandTokenizer.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
+package codeu.chat.util;
+
+import java.util.ArrayList;
+import java.util.List;
+
+final public class CommandTokenizer {
```
Comment (KevinWorkman) [inline src/codeu/chat/util/CommandTokenizer.java]: I would add some JavaDoc comments explaining how to use this class.

File: src/codeu/chat/util/CommandTokenizer.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+	private String source;
```
Comment (KevinWorkman) [inline src/codeu/chat/util/CommandTokenizer.java]: Similarly, comments explaining what these variables are used for would be helpful.

File: src/codeu/chat/util/CommandTokenizer.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	private int currentIndex = 0;
+
+	public boolean hasNextArg() {
+		int remainingIndices = source.length() - currentIndex;
+		return remainingIndices > 0;
+	}
+
+	// returns a String starting at currentIndex until the next occurrence of
+	// endingChar.
+	// Does not include endingChar in the String
```
Comment (KevinWorkman) [inline src/codeu/chat/util/CommandTokenizer.java:38]: Might want to also specify where it leaves `currentIndex`?

File: src/codeu/chat/util/CommandTokenizer.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		String toReturn = "";
+		while (hasNextArg() && source.charAt(currentIndex) != endingChar) {
+			toReturn += source.charAt(currentIndex);
+			currentIndex++;
+		}
+		currentIndex++;
+		return toReturn;
+	}
+
+	public String getNextArg() {
```
Comment (KevinWorkman) [inline src/codeu/chat/util/CommandTokenizer.java]: I'd also add comments on these functions that explain what they do and what state they leave the class in after they're called. As a general rule of thumb, I'd add JavaDoc to every public function and constructor.

**Review by KevinWorkman — COMMENTED** [review]
Awesome work! I don't have much to say here, other than I'd maybe add some more comments, especially on new classes and functions you create.

Also, make sure you sync up with Priyanka and Julia to make sure that their new commands work with your new tokenizer!

File: src/codeu/chat/util/CommandTokenizer.java (hunk 5 above)
Comment (TheMkrage) [reply src/codeu/chat/util/CommandTokenizer.java]: Should I use JavaDoc for these variables or just normal comments? I know JavaDoc is used by IDEs to provide method and variable description, but does that mean there is any disadvantage to using JavaDoc with private variables and methods?

Comment (KevinWorkman) [reply src/codeu/chat/util/CommandTokenizer.java]: No, there isn't really a disadvantage. If you have a ton of similar variables you probably don't need to JavaDoc them all (otherwise it would get pretty messy), but in this case since you only have two I think that would be absolutely fine.

File: src/codeu/chat/util/CommandTokenizer.java (hunk 7 above)
Comment (TheMkrage) [reply src/codeu/chat/util/CommandTokenizer.java]: Added JavaDoc to all public items (and some private )

File: test/codeu/chat/util/CommandTokenizerTest.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class CommandTokenizerTest {
+
+	@Test
+	public void testTokenizer() {
+		CommandTokenizer quotesTokenizer = new CommandTokenizer(
+				"Command \"arg0\" \"arg1\" \"arg2\" \"arg3\"");
```
Comment (KevinWorkman) [inline test/codeu/chat/util/CommandTokenizerTest.java:12]: Might want to include a test for this case, where two arguments have quotation marks but not spaces?

`Command "arg1""arg2"`

What is it supposed to do in this case? Adding a test will make it more obvious.

Comment (KevinWorkman) [reply test/codeu/chat/util/CommandTokenizerTest.java:12]: Similarly, you might want to add a test for when arguments inside quotation marks contain spaces.

**Review by KevinWorkman — COMMENTED** [review]
I'd say you're almost done! One small thing: I'd probably add a test for the case where arguments inside quotation marks contain spaces and you mix quotation marks and non-quotation marks, like this:

`command "arg one" argTwo argThree`

Similarly, maybe a test for issuing a command with no arguments? Just trying to think of corner cases that might break the code.

**Review by KevinWorkman — APPROVED** [review]
LGTM :tada:

One final note: I'd probably split your tests up into multiple functions, where each test function only tests one thing. Basically each of your `CommandTokenizer` and associated tests could be in their own function.

But I'd say this is ready to merge into the master branch! Like we discussed, you'll want to coordinate with Priyanka and Julia to make sure all your code is compatible.

**Review by Priya411 — APPROVED** [review]
